### PR TITLE
[SYCL][COMPAT] Move clashing functions into anonymous namespace

### PR DESCRIPTION
**Pull request by joeatodd**
Some `syclcompat::` functions clash with global namespace (C-style) functions because of Argument Dependent Lookup.

These are:
 - `memcpy`
 - `memset`
 - `free`
 - `fill`

To prevent ADL from finding these functions, they have been moved into an anonymous namespace.

### CODE REVIEW

**Review by Alcpz — APPROVED** [review]

File: sycl/include/syclcompat/memory.hpp [hunk 1]
```diff
@@ -644,6 +644,9 @@ static inline void wait_and_free(void *ptr,
   }
 }
 
+// Anonymous namespace to disable ADL for functions which might clash (memcpy,
```
Comment (ziranzha) [inline sycl/include/syclcompat/memory.hpp:647]: Will it be better to use an inline namespace?

Comment (joeatodd) [reply sycl/include/syclcompat/memory.hpp:647]: Hey @ziranzha, as far as I can tell, an  inline namespace doesn't fix the ADL issue. Here's a [godbolt](https://godbolt.org/z/1h94YhG8W) with anonymous namespace. If we replace it with an inline namespace, [it breaks](https://godbolt.org/z/7Wffxr5Y3) with the "ambiguous" error. Please let me know if I've misunderstood your suggestion. Thanks!

Comment (ziranzha) [reply sycl/include/syclcompat/memory.hpp:647]: LGTM

**Comment by joeatodd** [comment]
@intel/llvm-gatekeepers this is ready to merge :pray:

**Pull request merged**